### DISCUSSION TITLE
Detailed Error messages

### DISCUSSION THREAD

**Discussion by Cr0aten** [PHPMailer/PHPMailer#2696] [Q&A]
Hello,

Is there any way to get the detailed SMTP error message on failure.  For example, email providers such as AWS SES provide different error messages for different purposes.  If you try to send a message from an unauthorized address the PHPMailer debug info displays the error message as "554 Message rejected: Email address is not verified. The following identities failed the check in region US-EAST-2:", but the error message returned from the PHPMailer exception handler is "SMTP Error: data not accepted".  

Likewise, if you are sending messages too fast for your quota in SES, the PHPMailer debug info reports the error message as "554 Sending rate exceeded", but the error message returned from the PHPMailer exception handler is still "SMTP Error: data not accepted".  Both of the conditions each require completely different handling, but PHPMailer returns the same generic error instead of the detailed error when viewing debug info.

Thanks for the help!

**Comment by Synchro** [comment]
Unfortunately the SMTP spec doesn't define what appears after the error code, so in a generic SMTP client, it can only rely on the `554` code. You can get a more detailed error response with `getError()` on the SMTP instance, which will include what comes after the code, but it's up to you to parse it. Alternatively you can inject a debug handler and do it from there.

**Reply by Cr0aten** [reply]
Hi,

Thank you so much for responding so quickly.  I have tried pulling the information directly from getError(), but it returns the exact same information as the exception handler.  I do apologize but I don't fully understand why the detailed information can't be returned as part of the error message.  I do understand that only expected values, such as the error code, can be parsed for further processing, but if the server is returning a detailed description as well, wouldn't it be possible to also return that as part of the error message?

Thanks again!

**Reply by Synchro** [reply]
Well you can see in [the command handler that records the messages](https://github.com/PHPMailer/PHPMailer/blob/master/src/SMTP.php#L1006) that all it does is spot the error code and then stick whatever comes after it in the `details` property of the error info, so I'm not sure why you're not seeing that.

**Reply by Cr0aten** [reply]
Thanks again for replying so quickly!  I've been doing some additional testing to try and figure out why I'm not seeing the additional info in ErrorInfo.

When I call send() with SMTPKeepAlive **disabled**, the debug information looks like the snippet below and the contents of ErrorInfo **does include** the full text of the error message returned from the server.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
2022-06-11 23:20:57 CLIENT -> SERVER: .
2022-06-11 23:20:57 SERVER -> CLIENT: 554 Message rejected: Email address is not verified. The following identities failed the check in region US-EAST-2: test@sender.com
2022-06-11 23:20:57 SMTP ERROR: DATA END command failed: 554 Message rejected: Email address is not verified. The following identities failed the check in region US-EAST-2: test@sender.com
SMTP Error: data not accepted.
2022-06-11 23:20:57 CLIENT -> SERVER: QUIT
2022-06-11 23:20:57 SERVER -> CLIENT: 221 Bye
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

**ErrorInfo Contents:** SMTP Error: data not accepted.SMTP server error: DATA END command failed Detail: Message rejected: Email address is not verified. The following identities failed the check in region US-EAST-2: test@sender.com SMTP code: 554



However, when I call send() with SMTPKeepAlive **enabled**, the debug information looks like the snippet below and the contents of ErrorInfo **does not include** the full text of the error message returned from the server.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
2022-06-11 23:30:05 CLIENT -> SERVER: .
2022-06-11 23:30:05 SERVER -> CLIENT: 554 Message rejected: Email address is not verified. The following identities failed the check in region US-EAST-2: test@sender.com
2022-06-11 23:30:05 SMTP ERROR: DATA END command failed: 554 Message rejected: Email address is not verified. The following identities failed the check in region US-EAST-2: test@sender.com
SMTP Error: data not accepted.
2022-06-11 23:30:05 CLIENT -> SERVER: RSET
2022-06-11 23:30:05 SERVER -> CLIENT: 250 Ok
2022-06-11 23:30:05 CLIENT -> SERVER: QUIT
2022-06-11 23:30:05 SERVER -> CLIENT: 221 Bye
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

**ErrorInfo Contents:** SMTP Error: data not accepted.



So, it looks like when the call to smtp->reset is made to reset the connection and prepare for the next message (RSET) the contents of the  detailed error message is removed because there was no error during the call to smtp->reset.  

Does that sound correct or am I just doing something wrong?  Is there any way to still get the content of the detailed error message when SMTPKeepAlive is enabled?

Thanks.

**Reply by Synchro** [reply]
That sounds pretty likely. I'm not sure of the best way to fix that. Perhaps move issuing the RSET to the start of the next message rather than the end of this one, so that it remains available when you go around your sending loop?

**Reply by Cr0aten** [reply]
I replaced the code to clear the error in [line 1040](https://github.com/PHPMailer/PHPMailer/blob/master/src/SMTP.php#L1040):

$this->setError('');

with a conditional to skip clearing when running the 'RSET' command:

if($command != 'RSET') {
            $this->setError('');
        }

and it seems to do the trick.  I'm not sure if this would be the best way to do it, but it is a simple change and leaves the full content of any error message in tact.  I also can't see any negative impact when the loop starts over to process the next message.

Do you see any problems with doing it in this manner or have any suggestions for a better way to do it?

**Reply by Synchro** [reply]
The issue here is that then it's not possible to get an error response from the `RSET`, though to be fair that's very unlikely to happen! One place I can see that this might be a problem is if you are sending through a server that has a limit on the number of messages you can send during one connection session – it might fail the `RSET`, though I would expect it would be more likely to fail on the next `RCPT TO` or `DATA` command.

**Reply by Cr0aten** [reply]
I agree that an error on RSET would be pretty unlikely.  Event, if there was an error on RSET, the conditional wouldn't prevent that from logging the error...it just wouldn't clear it out after logging it.  Plus, an error on RSET would basically terminate the connection anyway...correct?  So, the close() at the end of the connection would call $this->setError('') anyway and then the next call to send() would establish a new connection and resume...at least that's the way it looks to me.
